### PR TITLE
containerd integration: produce progress events polling ctrd's content.Store

### PR DESCRIPTION
**Pull request by thaJeztah**
- upstreaming https://github.com/rumpl/moby/pull/12

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/containerd/service.go [hunk 1]
```diff
@@ -62,7 +63,38 @@ func (cs *ImageService) PullImage(ctx context.Context, image, tagOrDigest string
...
+		showProgress(ctx, jobs, cs.client.ContentStore(), outStream, stop)
+		stop <- struct{}{}
+	}()
+
+	img, err := cs.client.Pull(ctx, ref.String(), opts...)
+	if err != nil {
+		return err
+	}
+
+	unpacked, err := img.IsUnpacked(ctx, containerd.DefaultSnapshotter)
```
Comment (AkihiroSuda) [inline daemon/containerd/service.go:86]: (the snapshotter has to be eventually configurable, in a follow-up PR)

File: daemon/containerd/service.go [hunk 2]
```diff
@@ -62,7 +63,38 @@ func (cs *ImageService) PullImage(ctx context.Context, image, tagOrDigest string
 		}
 	}
 
-	_, err = cs.client.Pull(ctx, ref.String(), opts...)
+	jobs := newJobs()
+	h := containerdimages.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		if desc.MediaType != containerdimages.MediaTypeDockerSchema1Manifest {
```
Comment (AkihiroSuda) [inline daemon/containerd/service.go:68]: Should we return an error for schema1?

File: daemon/containerd/image_pull.go [hunk 3]
```diff
@@ -41,7 +42,38 @@ func (i *ImageService) PullImage(ctx context.Context, image, tagOrDigest string,
...
+	if err != nil {
+		return err
+	}
+
+	if !unpacked {
+		if err := img.Unpack(ctx, containerd.DefaultSnapshotter); err != nil {
+			return err
+		}
+	}
+	stop <- struct{}{}
```
Comment (cpuguy83) [inline daemon/containerd/image_pull.go:75]: What's the purpose of this? Seems like this races with your goroutine above.

Comment (cpuguy83) [reply daemon/containerd/image_pull.go:75]: Seems like we can close the channel (from the goroutine) rather than sending?

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: daemon/containerd/image_pull.go (hunk 3 above)
Comment (ndeloof) [reply daemon/containerd/image_pull.go:75]: closing channel would be a valid alternative, indeed

**Comment by thaJeztah** [comment]
Discussing; we can include https://github.com/rumpl/moby/pull/16 here, but also need to have a look at https://github.com/moby/moby/blob/f0680670d2303418816854a74e75e33e41907199/plugin/fetch_linux.go#L198-L285 to see if we can share the existing  implementation

**Comment by rumpl** [comment]
This one can be closed, superseeded by https://github.com/moby/moby/pull/44756

**Pull request closed without merging**